lib/spack/spack/database.py

```diff
@@ -0,0 +1,290 @@
+        """
+        if not self.is_dirty():
+            return
+
+        lock=0
+        while lock==0:
+            try:
+                os.mkdir(self.lock_path)
+                lock=1
+            except OSError as err:
+                pass
+
+        #The try statement ensures that a failure won't leave the
+        #database locked to other processes.
+        try:
+            if os.path.isfile(self.file_path):
+                with open(self.file_path,'r') as f:
+                    self.from_yaml(f)
+            else:
+            #The file doesn't exist, construct empty data.
+                self.data = []
+        except:
+            os.rmdir(self.lock_path)
+            raise
+
+        os.rmdir(self.lock_path)
+
+
+    def write_database_to_yaml(self,stream):
+        """
+        Replace each spec with its dict-node form
+        Then stream all data to YAML
+        """
+        node_list = []
+        spec_list = [sph['spec'] for sph in self.data]
+
+        for sph in self.data:
+            node = {}
+            deps = []
+            for name,spec in sph['spec'].dependencies.items():
+                deps.append(spec_list.index(spec))
+            node['spec']=Spec.to_node_dict(sph['spec'])
+            node['hash']=sph['hash']
+            node['path']=sph['path']
+            node['dependency_indices']=deps
+            node_list.append(node)
+
+        node_dict = dict(enumerate(node_list))
+        return yaml.dump({ 'database' : node_dict},
+                         stream=stream, default_flow_style=False)
+
+
+    def write(self):
+        """
+        Write the database to the standard location
+        Implements mkdir locking for the database file
+        """
+        lock=0
+        while lock==0:
+            try:
+                os.mkdir(self.lock_path)
+                lock=1
+            except OSError as err:
+                pass
+
+        #The try statement ensures that a failure won't leave the
+        #database locked to other processes.
+        try:
+            with open(self.file_path,'w') as f:
+                self.last_write_time = int(time.time())
+                self.write_database_to_yaml(f)
+        except:
+            os.rmdir(self.lock_path)
+            raise
+
+        os.rmdir(self.lock_path)
+
+
+    def get_index_of(self, spec):
+        """
+        Returns the index of a spec in the database
+        If unable to find the spec it returns -1
+        """
+        for index, sph in enumerate(self.data):
+            if sph['spec'] == spec:
+                return index
+        return -1
+
+
+    def is_dirty(self):
+        """
+        Returns true iff the database file exists
+        and was most recently written to by another spack instance.
+        """
+        return (os.path.isfile(self.file_path) and (os.path.getmtime(self.file_path) > self.last_write_time))
+
+
+#    @_autospec
+    def add(self, spec, path):
```

Comment:
I suspect that for concurrent edits, you want to modify this thing to keep track of the changes since last read, so that you can merge them with other spack instances' changes.